pkg/operator/controller/controller_dns.go

```diff
@@ -0,0 +1,82 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	ingressv1alpha1 "github.com/openshift/cluster-ingress-operator/pkg/apis/ingress/v1alpha1"
+	"github.com/openshift/cluster-ingress-operator/pkg/dns"
+
+	corev1 "k8s.io/api/core/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+// ensureDNS will create DNS records for the given LB service. If service is
+// nil, nothing is done.
+func (r *reconciler) ensureDNS(ci *ingressv1alpha1.ClusterIngress, dnsConfig *configv1.DNS, service *corev1.Service) error {
+	desiredDNSRecords, err := desiredDNSRecords(ci, dnsConfig, service)
+	if err != nil {
+		return err
+	}
+	for _, record := range desiredDNSRecords {
+		err := r.DNSManager.Ensure(record)
+		if err != nil {
+			return fmt.Errorf("failed to ensure DNS record %v for %s: %v", record, ci.Name, err)
+		}
+		logrus.Infof("ensured DNS record for clusteringress %s: %v", ci.Name, record)
+	}
+	return nil
+}
+
+// desiredDNSRecords will return any necessary DNS records for the given inputs.
+// If service is nil, no records are desired. If an ingress domain is in use,
+// records are desired in every specified zone present in the cluster DNS
+// configuration.
+func desiredDNSRecords(ci *ingressv1alpha1.ClusterIngress, dnsConfig *configv1.DNS, service *corev1.Service) ([]*dns.Record, error) {
+	records := []*dns.Record{}
+
+	// If no service exists, no DNS records should exist.
+	if service == nil {
+		return records, nil
+	}
+
+	// TODO: This will need revisited when we stop defaulting .spec.ingressDomain
+	// and .spec.highAvailability as both can be nil but used with an effective
+	// system-provided default reported on status.
+	if ci.Spec.HighAvailability.Type != ingressv1alpha1.CloudClusterIngressHA || ci.Spec.IngressDomain == nil {
```

Comment:
We do indeed want eventually to replace `ci.Spec.HighAvailability.Type` with an effective value, but as long as we are using `ci.Spec.HighAvailability`, we should still check whether it is nil.

Reply:
Fixed